Add ScrollView native component with default offset scrolling functionality

pprabhu-ng opened the pull request:
- Add some important loggings required for debugging
- Add stub functions for Native modules (Keyboard) required by ScrollView component 
- Define RN version and provide the same in Platform module
- Add ScrollView and ScrollContentView props registration in Legacy UI Manager
- Update compositor beginClip() to use paintContext damageRect & canvas values
- Add ScrollView and ScrollViewProvider 
- Add ScrollLayer in rns_shell - layer to handle scrolling bitmap 
- Simple Test apps used for vertical and horizontal scrolling tests

shefayaparwin commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,276 @@
+    forceBitmapReset_ = true;
+
+    Rect childFrame = newChildComponent->getComponentData().layoutMetrics.frame;
+    if((childFrame.origin.y + childFrame.size.height) > containerHeight_)
```
Is that the calculation intended to exclude/reduce the Edge inset set. If then won't it be  minus  contentInset value.
If I understood correctly, getContentFrame(), will this helps ?

pprabhu-ng replied:
this logic is to get the container width & height .
We have two frames here 
1) ScrollView frame  : which is visible area for displaying scrolling content. This is received from the layout metrics frame. 
2) Scrolling Content frame : which is the container where the content is available. This value will vary based on how many childrens we have . The calculation here is to get this value.

shefayaparwin replied:
Thanks for the detailed explanation. The need for adding  contentInset_ is where difficult for me to understand

shefayaparwin commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,276 @@
+
+    Rect childFrame = oldChildComponent->getComponentData().layoutMetrics.frame;
+    if(!isHorizontalScroll()) {
+       if(childFrame.origin.y < containerHeight_)
```
If the scroll content are placed with spacing using margin or padding. then resetting the height to the teh removed component origin will leave some space between the last element & the end of container. Is that can cause any impact .[As i understand container Height is used for allocating Bitmap & decide of scroll offset, will there be any case, which demands exact  container dimension ]

shefayaparwin replied:
Having a question ? Can a element in the middle of the scroll view container be removed. If possible, will the container Height won't be adjusted in this case.

pprabhu-ng replied:
Good point 👍 Yes this is possible. This logic will have issue in this case. 
Will update it to handle this as well.

shefayaparwin commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,276 @@
+  Rect frame = component.layoutMetrics.frame;
+  if(childList_.size()) {
+     if(scrollBitmap_.drawsNothing()) {
+        scrollBitmap_.allocN32Pixels(containerWidth_,containerHeight_);
```
When will this memory be freed

pprabhu-ng replied:
The bitmap memory would be available until the scroll view component is deleted by Delete instruction from ShadowView Mutation.
SkBitmap has the ownership of the pointer which points to this pixel memory. This pointer is a shared pointer.When the RSKcomponentScrollView object is destroyed, SKBitmap object is destroyed as well which reduces ref count of the smart pointer  and memory associated with the same is released.

pprabhu-ng commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,304 @@
+         Rect childFrame = child->getComponentData().layoutMetrics.frame;
+
+         /* Consider the last element in y direction to get the container height */
+         if((childFrame.origin.y + childFrame.size.height) > frameSize.height)
```
Use Rect getMaxX or getMaxY functions instead of (childFrame.origin.y + childFrame.size.height)

munezbn commented on ReactSkia/MountingManager.cpp:
```diff
@@ -66,6 +71,18 @@ void MountingManager::ProcessMutations(
+    RNS_LOG_DEBUG("\tTag:" << mutation.newChildShadowView.tag);
+    RNS_LOG_DEBUG("\tName:" << (mutation.newChildShadowView.componentName?mutation.newChildShadowView.componentName : "null"));
+    RNS_LOG_DEBUG("Mutation index : "<< mutation.index);
+
```
I think it should be ok to use single RNS_LOG_DEBUG here. Else if we want to enable only this part ( by changing to RNS_LOG_INFO) for quick debug, then we have to change each one of them to INFO. May be Add between two ==== lines.

pprabhu-ng replied:
Updated to a single RNS_LOG_INFO message

munezbn commented on ReactSkia/modules/RSkKeyboardObserver.h:
```diff
@@ -0,0 +1,28 @@
+  void startObserving() {};
+
+  void stopObserving() {};
+
```
Better we add Not implemented print here if it is not called too frequently

munezbn commented on ReactSkia/core_modules/RSkSpatialNavigator.cpp:
```diff
@@ -89,14 +91,15 @@ void RSkSpatialNavigator::removeFromNavList(std::shared_ptr<RSkComponent> candid
     if (it != navComponentList_.end()) {
         // Candidate found in the navigatable list, Reset data if candidate's focusable props have changed
-        if (viewProps.isTVSelectable == false && viewProps.focusable == false) {
+        if (viewProps.isTVSelectable == false && viewProps.focusable == false && !needEvent) {
```
We will be replacing this with scrollable property check right ?

pprabhu-ng replied:
update to use scrollEnabled value instead of needEvent call

munezbn commented on ReactSkia/MountingManager.cpp:
```diff
@@ -105,7 +122,7 @@ void MountingManager::CreateMountInstruction(
   if (provider) {
     std::shared_ptr<RSkComponent> component =
-        provider->CreateAndAddComponent(mutation.newChildShadowView);
+        provider->CreateAndAddComponent(mutation.newChildShadowView,surface_);
```
Will need to work on a different way to schedule display update on compositor of RNS SHELL

munezbn commented on ReactSkia/modules/RSkKeyboardObserver.cpp:
```diff
@@ -0,0 +1,20 @@
+  std::shared_ptr<CallInvoker> jsInvoker,
+  Instance *bridgeInstance)
+  : RSkEventEmitter(name, jsInvoker, bridgeInstance){}
+
```
If this module has methods better to add skeleton with NOT_IMPL macro.

pprabhu-ng replied:
This module has only two methods - add & remove listener, which is mapped in RSkEventEmitter.

munezbn commented on ReactSkia/components/RSkComponentScrollView.h:
```diff
@@ -0,0 +1,85 @@
+    const ShadowView &newShadowView,
+    bool forceUpadate) override;
+
+  bool needEvent() override;
```
This will be replaced with property.

munezbn commented on ReactSkia/components/RSkComponentScrollView.h:
```diff
@@ -0,0 +1,85 @@
+ private:
+  sk_sp<SkPicture> getPicture();
+
+  bool isHorizontalScroll();
```
Does RN has component with both Horizontal and Vertical scroll behavior enabled ? If no then fine.

pprabhu-ng replied:
No. Only one behavior can be enabled at a time. Either horizontal or vertical.

munezbn commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,313 @@
+
+using namespace RSkDrawUtils;
+
+RSkComponentScrollView::RSkComponentScrollView(const ShadowView &shadowView , RSkSurfaceWindow * surface)
```
Will  have online review for this file before adding further comments.

pprabhu-ng replied:
Resolved the comments identified in the discussion

munezbn commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,313 @@
+   Size frameSize = getComponentData().layoutMetrics.frame.size;
+   if(childList_.size()) {
+      for (auto &child : childList_) {
+         Rect childFrame = child->getComponentData().layoutMetrics.frame;
```
Need to consider absoluteFrame ( because of transformation)

munezbn commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,313 @@
+
+         /* Consider the last element in y direction to get the container height */
+         if((childFrame.getMaxY()) > frameSize.height)
+            frameSize.height = childFrame.getMaxY() + scrollContentInset_.bottom;
```
You should not consider scrollContentInset_.bottom here. It should be added at the end once we have traversed all the child

munezbn commented on ReactSkia/components/RSkComponentScrollView.cpp:
```diff
@@ -0,0 +1,313 @@
+
+         /* Consider the last element in x direction to get the container width */
+         if(childFrame.getMaxX() > frameSize.width)
+            frameSize.width = childFrame.getMaxX() + scrollContentInset_.right;
```
Same as above comment

pprabhu-ng replied:
This function is now required. We can get the content size from the ShadowView state object itself. So updated the code to use this value instead of calculation

pprabhu-ng committed:
Add Scroll Layer in rns_shell

Scroll Layer
1. Manages scrollable bitmap draw on canvas
2. Paint logic involves paint children and paint self
3. Provides means to set bitmap size , scrolling offset, shadow picture and border picture objects

pprabhu-ng committed:
Add ScrollView native component

1. Add scrollview native component handling manual scrolling with offset=screensize/2
2. Register component provider and props,add stub native module required
3. Cosmetic updates related to header file includes or function definition
4. Add some importing loggings
5. Update proper RN version

pprabhu-ng committed:
Scroll View manual scrolling test apps

Test apps to verify different use cases of manual scrolling

pprabhu-ng committed:
Resolve build issue

pprabhu-ng committed:
update comments

pprabhu-ng committed:
Update required invalidate mask during invalidate

munezbn commented on ReactSkia/components/RSkComponent.h:
```diff
@@ -1,20 +1,26 @@
+   PictureTypeBorder = 1 << 0,
+   PictureTypeAll = 1 << 1,
+};
+
```
II would suggest start with 1<< 0 and All = basically OR with remaining

munezbn commented on ReactSkia/core_modules/RSkSpatialNavigator.cpp:
```diff
@@ -54,7 +54,7 @@ void RSkSpatialNavigator::addToNavList(std::shared_ptr<RSkComponent> newCandidat
                                     << viewProps.isTVSelectable << "]" << " focusable[" << viewProps.focusable << "] " <<
                                     "hasTVPreferredFocus[" << viewProps.hasTVPreferredFocus << "]" << " accessible[" << viewProps.accessible << "]");
-    if (viewProps.isTVSelectable == true || viewProps.focusable == true) {
+    if (viewProps.isTVSelectable == true || viewProps.focusable == true | newCandData.commonProps.scrollEnabled == true) {
```
Or symbol is wrong

munezbn commented on ReactSkia/components/RSkComponentScrollView.h:
```diff
@@ -0,0 +1,56 @@
+enum ScrollDirectionMask {
+   ScrollDirectionForward = 0,
+   ScrollDirectionBackward = 1 << 0,
+};
```
Same comment as above.

pprabhu-ng committed:
addressed review comments

pprabhu-ng committed:
address cosmetic issues

pprabhu-ng committed:
Resolve conflicts with latest base
